framework/include/base/MooseServer.h

```diff
@@ -0,0 +1,172 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifdef WASP_ENABLED
+
+#pragma once
+
+#include <string>
+#include <memory>
+#include "wasplsp/LSP.h"
+#include "wasplsp/ServerImpl.h"
+#include "wasplsp/Connection.h"
+#include "wasplsp/IOStreamConnection.h"
+#include "waspcore/Object.h"
+#include "wasphit/HITNodeView.h"
+#include "MooseApp.h"
+
+class MooseServer : public wasp::lsp::ServerImpl
+{
+public:
+  MooseServer() : _is_setup(false)
+  {
+    // set server connection to new iostream connection shared pointer
+
+    _connection = std::make_shared<wasp::lsp::IOStreamConnection>(this);
+
+    // set server capabilities to receive full input text when changed
+
+    server_capabilities[wasp::lsp::m_text_doc_sync] = wasp::DataObject();
+    server_capabilities[wasp::lsp::m_text_doc_sync][wasp::lsp::m_open_close] = true;
+    server_capabilities[wasp::lsp::m_text_doc_sync][wasp::lsp::m_change] = wasp::lsp::m_change_full;
+  }
+
+  /** get read / write connection - specific to this server implemention
+   * @return - shared pointer to the server's read / write connection
+   */
+  std::shared_ptr<wasp::lsp::Connection> getConnection() { return _connection; }
+
+  /** setup server for input validation and template based autocompletion
+   * @param moose_app - pointer to moose app calling this set up method
+   * @return - true if server setup was successful and everything was set
+   */
+  bool setup(MooseApp * moose_app);
```

Comment:
Why is this method necessary? It looks you should store a reference to the MooseApp instead as a member variable, and pass the app in the constructor of the MooseServer.